### ISSUE TITLE
Batch norm individual gradients

### ISSUE THREAD

**Issue by schaefertim** [f-dangel/backpack#160]
`BatchNorm` is a special module because it mixes samples within the batch. This needs some special treatment.

PyTorch documentation: [BatchNorm1d](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html), [BatchNorm2d](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html)

If `BatchNorm` is in evaluation mode (`.training=False`), the saved statistics are used. This is independent of the batch. Therefore, individual gradients are well defined.

However, if `BatchNorm` is in training mode (`.training=True`), the batch statistics are used instead. Therefore, individual gradients are not well defined anymore. I suggest multiple possibilities of realizing individual gradients:
1. Use the saved statistics like in evaluation mode.
2. Use `batch_size=1`, i.e. E(x)=x and Var(x)=0.
3. Use complete batch statistics. This is the only case where `.grad_batch.sum(axis=0)==.grad` holds.
4. Use complete batch statistics and approximate individual gradients by neglecting the mixed terms. This leads to separate individual gradients.

I favor the third alternative. If the other approaches do have some merit, it is also possible to implement a switch. For example requesting `module.batch_norm_mode: str` and execute the given mode if provided.

Note: The [current version](https://github.com/f-dangel/backpack/blob/d17408120c30e21bbc7d038fb10b680375940939/backpack/core/derivatives/batchnorm1d.py#L18) does not bother about different modes. I think it exclusively implements the third alternative, but I need some time to verify this.

**Comment by f-dangel** [comment]
More details on this topic can be found [here](https://f-dangel.github.io/about/#filepostsbatchnormalizationorgStructuralimplicationsofbatchnormalization).

**Comment by f-dangel** [comment]
Basic support added in `1.4.0`, see the [changelog](https://github.com/f-dangel/backpack/blob/master/changelog.md#140---2021-11-12).

Closing.
